lib/src/support/canvas.cpp

```diff
@@ -247,6 +247,27 @@ namespace cycfi { namespace elements
       cairo_close_path(&_context);
    }
 
+   void canvas::add_round_rect(struct rect r_, std::array<float, 4> radius)
+   {
+      auto x = r_.left;
+      auto y = r_.top;
+      auto r = r_.right;
+      auto b = r_.bottom;
+      auto const a = M_PI/180.0;
+      for (auto& r : radius) {
+         r = std::min(r, std::min(r_.width(), r_.height()) / 2);
+      }
+
+      // radii array follow the unit circle defined counterclockwise with 0 at (x, y) = (1, 0)
+      // we have to map this to the cairo definition
+      cairo_new_sub_path(&_context);
```

Comment:
This is Cairo specific. There should be equivalent canvas functions so it can port to skia directly.

Reply:
I just followed the implementation of the existing `add_round_rect` method. Should this be ported to skia or redefined in terms of existing methods? Considering this is a more general version of the existing method, I think it would make sense to just replace the old one with this.

Reply:
> I just followed the implementation of the existing `add_round_rect` method. Should this be ported to skia or redefined in terms of existing methods? Considering this is a more general version of the existing method, I think it would make sense to just replace the old one with this.

No. Skia and Quartz-2D already have round_rect. Cairo does not. What you are seeing is a workaround for the lack of native round_rect facility in Cairo. 

Elements itself (apart from canvas, which is ported to artist for different backends), should use only canvas facilities. have a look at the canvas arc functions: https://github.com/cycfi/elements/blob/master/lib/include/elements/support/canvas.hpp#L65-L71

Reply:
> I think it would make sense to just replace the old one with this.

Canvas, and hence Artist, follows the W3C canvas API. round rect is a specialized API which assumes a single radius for all corners. If you want something more specialized than that, then use primitive lines and arcs and even splines.

Reply:
In general, any new code or code change should also compile, build and run with the `skia_2024` branch. Please make sure of that.

Reply:
That makes sense. I didn't know the canvas api was fixed.

Reply:
> That makes sense. I didn't know the canvas api was fixed.

Yeah, sorry it was not documented (yet). Working on it.